picocrt/machine/arm/crt0.c

```diff
@@ -188,23 +188,28 @@ extern char __stack[];
 #define I_BIT           (1 << 7)
 #define F_BIT           (1 << 6)
 
+#define SET_MODE(mode) \
+    __asm__("mov r0, %0\nmsr cpsr_c, r0" :: "r" (mode | I_BIT | F_BIT): "r0")
+
 #define SET_SP(mode) \
-    __asm__("mov r0, %0\nmsr cpsr_c, r0" :: "r" (mode | I_BIT | F_BIT): "r0");   \
+    SET_MODE(mode);  \
     __asm__("mov sp, %0" : : "r" (__stack))
 
 #define SET_SPS()                               \
         SET_SP(MODE_IRQ);                       \
         SET_SP(MODE_ABT);                       \
         SET_SP(MODE_UND);                       \
         SET_SP(MODE_FIQ);                       \
-        SET_SP(MODE_SVC);                       \
```

Comment:
This is getting skipped on all other targets now. I think the right solution will be to always set the SVC SP in _start and then make SET_SPS always do SET_MODE(MODE_SVC) as the last operation, and not do SET_SP(MODE_SVC) at all.

Reply:
see my comment below

Reply:
My concern is that non Thumb-1 targets are going to be left in FIQ mode because they only use `SET_SPS` from `_start`. By ending `SET_SPS` with `SET_MODE(MODE_SVC)`, both users of that code end up with the same result.

Reply:
yes, agreed. taking the non-m-class targets in concideration, we should add that here, like you mentioned.

Reply:
and sorry for the delay in progressing this. should be back in operation later this week.

Reply:
> Shall I just merge my suggested change then? Or would you like something different?

i wanted the code to do the sensible thing for m-class cores, and not write the same address for each sp.

Reply:
I'm a bit confused -- I thought all of the m-class cores only had two stack pointers, Main and Process? All of the changes in this series are used when `_ARM_ARCH_PROFILE != 'M'`; did you have some plans for the 'M' code?

Reply:
Yes, your M-core description is also my understanding. But the patch you posted [still does](https://github.com/picolibc/picolibc/commit/4106a3e991d14a6f76d66510c9dd7d678b1509be#diff-684d9507a0223b37a330c8d1e2bb265ce3878e1e4502cf38e49f105be7bbe9aeL308) mode changes and sp writes for m-cores.

Reply:
I don't think it does? https://github.com/picolibc/picolibc/commit/4106a3e991d14a6f76d66510c9dd7d678b1509be should only change code in the '#else' block starting around line 96 and ending around line 315. Or is there some target you're thinking of which doesn't have `__ARM_ARCH_PROFILE == 'M'`?

Reply:
Ah right, that's what you're trying to say!
Sorry I missed that #if #else :(

Ok so yea, I guess then it's only preferring to not have every sp start with same address, but that feels more like an improvement. I got no issues with you pushing what you have.

Reply:
Thanks for your review. I'll push what I've got, but I'd love to see a PR fixing the stack sharing issue.